### PR TITLE
fix BlobStorage.read to initialize new blob if blob doesn't exist

### PR DESCRIPTION
**Pull request by stevengum**
Should address #314...

### CODE REVIEW

File: libraries/botbuilder-azure/src/blobStorage.ts [hunk 1]
```diff
@@ -117,14 +117,27 @@ export class BlobStorage implements Storage {
...
                 Promise.all<DocumentStoreItem>(sanitizedKeys.map((key) => {
-                    return new Promise((resolve, reject) => {
-                        this.client.getBlobMetadataAsync(container.name, key).then((blobMetadata) => {
-                            this.client.getBlobToTextAsync(blobMetadata.container, blobMetadata.name).then((result) => {
-                                let document: DocumentStoreItem = JSON.parse(result as any);
-                                document.document.eTag = blobMetadata.etag;
-                                resolve(document);
-                            }, err => resolve(null));
-                        }, (err) => resolve(null));
+                    return this.checkForBlob(container.name, key).then((blobResult) => {
```
Comment (johnataylor) [inline libraries/botbuilder-azure/src/blobStorage.ts]: no need to define a one line function here - just put the check inline (all the other code uses client)

File: libraries/botbuilder-azure/src/blobStorage.ts [hunk 2]
```diff
@@ -117,14 +117,27 @@ export class BlobStorage implements Storage {
...
+                        if (blobResult.exists) {
+                            return this.client.getBlobMetadataAsync(container.name, key).then((blobMetadata) => {
+                                return this.client.getBlobToTextAsync(blobMetadata.container, blobMetadata.name).then((result) => {
+                                    const document: DocumentStoreItem = JSON.parse(result as any);
+                                    return document;
```
Comment (johnataylor) [inline libraries/botbuilder-azure/src/blobStorage.ts]: don't we still need the eTag?

File: libraries/botbuilder-azure/src/blobStorage.ts [hunk 3]
```diff
@@ -222,6 +234,10 @@ export class BlobStorage implements Storage {
         return checkedCollections[key];
     }
 
+    private checkForBlob(containerName: string, blob: string): Promise<azure.BlobService.BlobResult> {
```
Comment (johnataylor) [inline libraries/botbuilder-azure/src/blobStorage.ts]: see previous comment - pulling this code into this function adds little value

File: libraries/botbuilder-azure/src/blobStorage.ts (hunk 2 above)
Comment (stevengum) [reply libraries/botbuilder-azure/src/blobStorage.ts]: This is the document from blobStorage, so it already has an eTag in the current implementation.

There is a related issue saying that BotState should just be POCO (#297); so in the code below should we be bothering with eTags at all?

File: libraries/botbuilder-azure/src/blobStorage.ts [hunk 4]
```diff
@@ -117,14 +117,27 @@ export class BlobStorage implements Storage {
...
+                                accessConditions: azure.AccessCondition.generateEmptyCondition(),
+                                parallelOperationThreadCount: 4
+                            } as azure.BlobService.CreateBlobRequestOptions;
+                            return this.client.createBlockBlobFromTextAsync(container.name, blobResult.name, '{}', createBlobOptions).then((creationResult) => {
+                                return this.client.getBlobToTextAsync(creationResult.container, creationResult.name).then((result) => {
+                                    const document: DocumentStoreItem = JSON.parse(result as any);
+                                    document.document = { eTag: '*' };
+                                    return document;
+                                });
+                            });
```
Comment (tomlm) [inline libraries/botbuilder-azure/src/blobStorage.ts]: Why are we doing this delta at all?  God sheds a tear when a blob is wasted.

File: libraries/botbuilder-azure/src/blobStorage.ts (hunk 2 above)
Comment (stevengum) [reply libraries/botbuilder-azure/src/blobStorage.ts]: The two (#297 and this) aren't mutually exclusive so I added the eTag semantics back to the PR. The PR is much simpler now than it was originally.

File: libraries/botbuilder-azure/src/blobStorage.ts (hunk 1 above)
Comment (stevengum) [reply libraries/botbuilder-azure/src/blobStorage.ts]: Done!

**Comment by stevengum** [comment]
The PR is much simpler now, and `.read()` will return an empty DocumentStoreItem if the blob does not exist. `BlobStorage.read()` no longer swallows Errors and the tests now show the actual Error.message instead of just displaying an Error object.

@johnataylor, @tomlm

**Pull request merged**